proposals/0153-elgamal-proof-program.md

```diff
@@ -0,0 +1,240 @@
+quite specialized for the SPL Token program. For example, the program contains
+instructions `VerifyTransfer`, `VerifyTransferWithFee`, and `VerifyWithdraw`
+that verify zero-knowledge proofs that are tailor made for `Transfer` and
+`Withdraw` instructions in the SPL Token confidential transfer extension.
+
+The ZK Token Proof program is a native built-in program that is part of the
+validator client. These type of programs should generally not favor any specific
+application/program like the SPL Token program. Furthermore the logic
+contained in the ZK Token Proof program enables private transfer of SPL tokens.
+The SPL Token program is limited to confidential transfers (not
+anonymous) and it contains an audit feature which enables authorities to decrypt
+any confidential transfers. Nevertheless, including these type of privacy related
+proof verification logic in the validator client can provide unnecessary legal
+burden to the maintainers of the Solana validator clients.
+
+In this document, we propose that we deprecate the existing ZK Token Proof
+program and replace it with a more general ZK ElGamal Proof program that is
+application independent. The new ZK ElGamal Proof program inherits parts of the
+ZK Token Proof program that is independent of any specific application like
+the logic to verify the validity of a public key or range of the encrypted in an
+ElGamal ciphertext. It leaves out parts of the logic that is specific
+to the SPL Token application like the logic to verify a zero-knowledge proof
+required for a token transfer instruction.
+
+## Alternatives Considered
+
+We can activate the ZK Token Proof program. However, as explained above, this is
+too specific to a particular application and also provides unnecessary burden to
+Solana validator clients as the program contains privacy related logic.
+
+We can deprecate the ZK Token Proof program. However, the program still contains
+useful zero-knowledge proofs that are very general and will benefit many
+projects in the ecosystem.
+
+We can rewrite/compile the logic in the ZK Token Proof program as BPF
+instructions, so that the program is independent of any validator client.
+Unfortunately, the logic in the proof program are just too expensive to be
+currently run inside the vm. This could change in the future; however,
+launching the ZK Token Proof (or the new ZK ElGamal Proof) is of high priority
+to the ecosystem as there are currently a number of projects that are waiting for
+the ZK Token Proof program to be activated.
+
+## New Terminology
+
+n/a
+
+## Detailed Design
+
+The new ZK ElGamal Proof program contains a strict subset of the proof
+verification instructions in the original ZK Token Proof program. We list the
+current list of the instructions in the ZK Token Proof program and denote
+whether they are included, removed, or renamed in the new ZK ElGamal Proof
+program.
+
+```rust
+pub enum ProofInstruction {
```

Comment:
nit: rather than using a rust enum and code, which can make it hard to parse quickly, can you simply list the instruction names and differences? ie

* `CloseContextState`: unchanged
* `VerifyZeroBalance`: renamed to `VerifyZeroCiphertext`

And if you want, you can provide a link to the current ZK Token Proof program instruction code.

Reply:
Okay good idea. I removed the rust enum code and just listed the instructions that are to be renamed or removed. I can even remove the short proof descriptions, but I thought maybe I can keep it just for context. I can remove it if you prefer removing it too.

Reply:
What you did looks great, thanks!